Fix client.Clone() to include the address.

raskchanky opened the pull request:
Additionally, add getters for any setters that didn't already have them. Also beef up the test to exhaustively test every getter/setter combo.

Closes https://github.com/hashicorp/vault/issues/9905

ncabatoff commented on api/client.go:
```diff
@@ -499,6 +502,15 @@ func (c *Client) SetLimiter(rateLimit float64, burst int) {
+	c.modifyLock.RLock()
+	c.config.modifyLock.RLock()
+	defer c.config.modifyLock.RUnlock()
+	c.modifyLock.RUnlock()
```
Not sure it's safe to unlock here, since we're referencing c.config afterwards.  Why not just defer for both locks?

raskchanky replied:
Sure, good call. I'm embarrassed to say I just copy/pasted this implementation from one of the existing getter functions without examining it carefully. I'll go through all of the getter functions and double check this.

ncabatoff commented on api/client.go:
```diff
@@ -670,24 +709,41 @@ func (c *Client) SetBackoff(backoff retryablehttp.Backoff) {
+// One exception to this is the addr field of Client, which is also copied over.
+// Otherwise you have the confusing result of creating a client, customizing
+// the address via client.SetAddress() then cloning it and having it show the
+// default address again, instead of the one you just set. I'm not even sure
```
I don't think this comment is the right place to pose such questions ("I'm not even sure...").  My guess is that yes, it's a convenience/optimization.

ncabatoff replied:
I'm not sure this comment update even needs to be added.  The address is part of the config, so the existing comment implies it should be in scope of Clone.  The fact that there's some additional work we need to do such that SetAddress is not necessary post-Clone isn't really relevant to the consumer of the doc for this method.  If you want to document why we do SetAddress, I'd put that as an inline comment above where we invoke it.

raskchanky replied:
Sounds good

ncabatoff commented on api/client.go:
```diff
@@ -670,24 +709,41 @@ func (c *Client) SetBackoff(backoff retryablehttp.Backoff) {
+// why the client has addr as a member at all, other than the convenience of
+// having it pre-parsed into a URL instead of a string.
 func (c *Client) Clone() (*Client, error) {
 	c.modifyLock.RLock()
```
Why not just grab both locks at the outset and defer release them?  They're just read locks, and this way the code will be easier to read and reason about.

raskchanky replied:
Sure

ncabatoff commented on api/client.go:
```diff
@@ -670,24 +709,41 @@ func (c *Client) SetBackoff(backoff retryablehttp.Backoff) {
+	}
 
-	return NewClient(newConfig)
+	err = client.SetAddress(newConfig.Address)
```
Looking at the code some more I'm confused why we need to call SetAddress here, since NewClient does the equivalent.

raskchanky replied:
I'll double check this to be sure, but my memory is that either `client.addr` or `config.Address` ended up not being set correctly without this line.

ncabatoff commented on api/client.go:
```diff
@@ -670,24 +709,41 @@ func (c *Client) SetBackoff(backoff retryablehttp.Backoff) {
-		Backoff:    config.Backoff,
-		CheckRetry: config.CheckRetry,
-		Limiter:    config.Limiter,
+		Address:          config.Address,
```
Want to add AgentAddress and SRVLookup to the copied fields while you're at it?

raskchanky replied:
Good call

raskchanky commented:
@ncabatoff I believe I addressed all of your feedback, in addition to adding 2 new getters for `AgentAddress` and `SRVLookup`, and some additional testing.

ncabatoff commented on api/client.go:
```diff
@@ -492,57 +495,129 @@ func (c *Client) Address() string {
+	return c.config.MaxRetries
+}
+
+func (c *Client) SetAgentAddress(address string) {
```
Can we do the same URL validation here as in SetAddress?

ncabatoff replied:
Actually I'm not sure we even want this method.  Right now AgentAddress is only used in NewClient, in which case it overrides any address configured.  It's weird to provide a setter that has no effect.

ncabatoff commented on api/client.go:
```diff
@@ -492,57 +495,129 @@ func (c *Client) Address() string {
+	c.config.AgentAddress = address
+}
+
+func (c *Client) AgentAddress() string {
```
Similarly to SetAgentAddress, I'm not sure what the use case for this is.

raskchanky replied:
Mostly I was just trying to be consistent in making sure everything had a getter/setter, but I can remove this and its corresponding setter.
